core/includes/install.core.inc

```diff
@@ -1454,6 +1450,372 @@ function install_select_language_form($form, &$form_state, $files) {
+ * English and for which there is no local translation file.
+ *
+ * @param string $langcode
+ *   Language code to check for download.
+ * @param string $server_pattern
+ *   Server access pattern (to replace language code, version number, etc. in).
+ *
+ * @return array
+ *   Requirements compliance array. If the translation cannot be downloaded this
+ *   will contain a requirements error with detailed information.
+ *
+ * @see install_state_defaults()
+ * @see install_download_translation()
+ */
+function install_check_translation_download($langcode, $server_pattern) {
+  $requirements = array();
+
+  $readable = FALSE;
+  $writable = FALSE;
+  $translation_available = FALSE;
+
+  $files_directory = conf_path() . '/files';
+  $translations_directory = settings_get('locale_translate_file_directory', conf_path() . '/files/translations');
+  $translations_directory_exists = FALSE;
+  $online = FALSE;
+
+  // First attempt to create or make writable the files directory.
+  file_prepare_directory($files_directory, FILE_CREATE_DIRECTORY | FILE_MODIFY_PERMISSIONS);
+
+  // Then, attempt to create or make writable the translations directory.
+  file_prepare_directory($translations_directory, FILE_CREATE_DIRECTORY | FILE_MODIFY_PERMISSIONS);
+
+  // Get values so the requirements errors can be specific.
+  if (backdrop_verify_install_file($translations_directory, FILE_EXIST, 'dir')) {
+    $readable = is_readable($translations_directory);
+    $writable = is_writable($translations_directory);
+    $translations_directory_exists = TRUE;
+  }
+
+  $version = BACKDROP_VERSION;
+
+  // For dev releases, remove the '-dev' part and trust the translation server
+  // to fall back to the latest stable release for that branch.
+  // @see locale_translation_build_projects()
+  if (preg_match("/^(\d+\.).*-(dev|preview)$/", $version, $matches)) {
+    // Example match: 1.28-dev => 1.x (Backdrop core).
+    $version = $matches[1] . 'x';
+  }
+
+  // Build URL for the translation file and the translation server.
+  $variables = array(
+    '%project' => 'backdropcms',
+    '%version' => $version,
+    '%core' => 'all',
+    '%language' => $langcode,
+  );
+  $translation_url = strtr($server_pattern, $variables);
+
+  $elements = parse_url($translation_url);
+  $server_url = $elements['scheme'] . '://' . $elements['host'];
+
+  // Build the language name for display.
+  include_once BACKDROP_ROOT . '/core/includes/standard.inc';
+
+  $languages = standard_language_list();
+  $language = isset($languages[$langcode]) ? $languages[$langcode][0] : $langcode;
+
+  // Check if any of the desired translation files are available or if the
+  // translation server can be reached. In other words, check if we are online
+  // and have an internet connection.
+  if (install_check_localization_server($server_url)) {
+    $online = TRUE;
+  }
+  if ($online) {
+    $translation_available = install_check_localization_server($translation_url);
```

Comment:
Something's weird with this logic... :thinking: 

The result of function install_check_localization_server() is statically cached, regardless of the URI, right?
So checking here twice, but with different URL doesn't seem to work as intended.
Additionally this function has already run in install_select_language().

So, is this an attempt to use install_check_localization_server() in a way, that's not intended?

Or am I missing something?

Reply:
The static check is saved only for if the server is **not** online (return is false) so it wont do another `http_request` on subsequent calls if the server is offline the first call. And it just checks for a response, so there shouldn't be any issue calling with different URL params.

So for this particular code block we check to see if the server is online, then we check if the translation exists. 

Of course, as we discussed before, we could easily skip both these checks, since we filter the language list with a similar check to get to this code. The hypothetical situation of the server going offline between selecting a language and clicking "Save and continue" isn't likely at all, and in either case doesnt work, as we noted.

Reply:
> The static check is saved only for if the server is not online

Ah, now I get it, right, that's what I missed.

> ... we could easily skip both these checks, since we filter the language list

Caution, there's no filter with commandline install, so checking makes sense for that type of install. The related error message shows up, if the language isn't available on the server. Better leave this check in.

This can be marked as "resolved". :+1: